liblangutil/SourceReferenceExtractor.cpp

```diff
@@ -0,0 +1,102 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <liblangutil/SourceReferenceExtractor.h>
+#include <liblangutil/CharStream.h>
+#include <liblangutil/Exceptions.h>
+
+#include <cmath>
+#include <iomanip>
+
+using namespace std;
+using namespace dev;
+using namespace langutil;
+
+namespace
+{
+struct LineColumn
+{
+	int line;
+	int column;
+
+	LineColumn(tuple<int, int> const& t): line{get<0>(t)}, column{get<1>(t)} {}
+};
+}
+
+SourceReferenceExtractor::Message SourceReferenceExtractor::extract(Exception const& _exception, string _category)
+{
+	SourceLocation const* location = boost::get_error_info<errinfo_sourceLocation>(_exception);
+
+	string const* description = boost::get_error_info<errinfo_comment>(_exception);
+	SourceReference primary = extract(location, description ? *description : "");
+
+	std::vector<SourceReference> secondary;
+	auto secondaryLocation = boost::get_error_info<errinfo_secondarySourceLocation>(_exception);
+	if (secondaryLocation && !secondaryLocation->infos.empty())
+		for (auto const& info: secondaryLocation->infos)
+			secondary.emplace_back(extract(&info.second, info.first));
+
+	return Message{std::move(primary), _category, std::move(secondary)};
+}
+
+SourceReference SourceReferenceExtractor::extract(SourceLocation const* _location, std::string description)
+{
+	if (!_location || !_location->source.get()) // Nothing we can extract here
+		return SourceReference{"", "", -1, -1, -1, -1, -1, -1, 0, std::move(description)};
+
+	auto const& source = _location->source;
+
+	LineColumn const interest = source->translatePositionToLineColumn(_location->start);
+	LineColumn start = interest;
```

Comment:
What is the difference between interest and start?

Reply:
(start, end) is the range that is potentially truncated, and thus modified (if truncated) whereas interest line/col is the actual position of "interest".

Reply:
I'd made that point (hopefully) clear in the `SourceReference` struct.